ttnn/cpp/ttnn/device_operation.hpp

```diff
@@ -0,0 +1,319 @@
+       const typename operation_t::tensor_args_t& tensor_args) {
+        operation_t::compute_program_hash(operation_attributes, tensor_args);
+    };
+};
+
+template <typename... Ts>
+[[nodiscard]] std::variant<Ts...> constexpr map_index_to_variant(std::size_t i, std::variant<Ts...>) {
+    assert(i < sizeof...(Ts));
+    static constexpr std::variant<Ts...> table[] = { Ts{ }... };
+    return table[i];
+}
+
+template <typename T>
+    requires std::same_as<std::decay_t<T>, Tensor>
+constexpr auto visit_tensor(auto callback, T&& value) {
+    callback(value);
+}
+
+template <typename T>
+constexpr auto visit_tensor(auto callback, const std::optional<T>& value) {
+    if (value.has_value()) {
+        const auto& tensor = value.value();
+        visit_tensor(callback, tensor);
+    }
+}
+
+template <typename T>
+constexpr auto visit_tensor(auto callback, const std::vector<T>& value) {
+    for (auto& tensor : value) {
+        visit_tensor(callback, tensor);
+    }
+}
+
+template <typename T, auto N>
+constexpr auto visit_tensor(auto callback, const std::array<T, N>& value) {
+    for (auto& tensor : value) {
+        visit_tensor(callback, tensor);
+    }
+}
+
+template <typename... Ts>
+constexpr auto visit_tensor(auto callback, const std::tuple<Ts...>& value) {
+    constexpr auto num_attributes = sizeof...(Ts);
+    [&callback, &value]<size_t... Ns>(std::index_sequence<Ns...>) {
+        (visit_tensor(callback, std::get<Ns>(value)), ...);
+    }(std::make_index_sequence<num_attributes>{});
+}
+
+template <typename T>
+    requires(not std::same_as<std::decay_t<T>, Tensor>) and requires { std::decay_t<T>::attribute_names; }
+constexpr auto visit_tensor(auto callback, T&& object) {
+    constexpr auto num_attributes = std::tuple_size_v<decltype(std::decay_t<T>::attribute_names)>;
+    visit_tensor(callback, object.attribute_values());
+}
+
+template <typename T>
+    requires std::same_as<std::decay_t<T>, Tensor>
+constexpr auto get_first_tensor(T&& value) {
+    return std::cref(value);
+}
+
+template <typename T>
+constexpr auto get_first_tensor(const std::optional<T>& value) {
+    if (value.has_value()) {
+        const auto& tensor = value.value();
+        return get_first_tensor(tensor);
+    }
+}
+
+template <typename T>
+constexpr auto get_first_tensor(const std::vector<T>& value) {
+    for (auto& tensor : value) {
+        return get_first_tensor(tensor);
+    }
+}
+
+template <typename T, auto N>
+constexpr auto get_first_tensor(const std::array<T, N>& value) {
+    for (auto& tensor : value) {
+        return get_first_tensor(tensor);
+    }
+}
+
+template <typename... Ts>
+constexpr auto get_first_tensor(const std::tuple<Ts...>& value) {
+    constexpr auto num_attributes = sizeof...(Ts);
+    return get_first_tensor(std::get<0>(value));
+}
+
+template <typename T>
+    requires requires { std::decay_t<T>::attribute_names; } and (not std::same_as<std::decay_t<T>, Tensor>)
```

Comment:
```suggestion
    requires { std::decay_t<T>::attribute_names; } and (not std::same_as<std::decay_t<T>, Tensor>)
```

Reply:
Consider removing the requires requires like this...
```
template <typename T>
concept HasAttributeNames = requires {
    typename std::decay_t<T>::attribute_names;
};

template <typename T>
concept NotTensor = not std::same_as<std::decay_t<T>, Tensor>;

template <typename T>
concept TensorArgs = HasAttributeNames<T> && NotTensor<T>;

template <TensorArgs T>
constexpr auto get_first_tensor(T&& object) {
...
}
```